lib/Parse/ParseStmt.cpp

```diff
@@ -588,7 +591,9 @@ ParserResult<Stmt> Parser::parseStmt() {
   if (isContextualYieldKeyword()) {
     Tok.setKind(tok::kw_yield);
   }
-  
+
+  // This needs to handle everything that `Parser::isStartOfStmt()` accepts as
```

Comment:
Can this be an assertion instead of a comment?

Reply:
Not sure how to do that.

Reply:
Added an assertion for `Parser::parseStmt()` making lexing progress before returning.

Reply:
Perhaps put an assertion in the `default` case that `isStartOfStmt()` is false? That would only be partially effective, though.

Reply:
> Perhaps put an assertion in the `default` case that `isStartOfStmt()` is false?

But that was true while the issue was present, otherwise `parseStmt()` would not be called in the first place.
The problem was that `isStartOfStmt()` was returning true when pointing at an `@` token at statement context but `parseStmt()` was not handling it in erroneous position.

I think the assertion I added is enough to make sure to crash with an assertion instead of going into infinite loop, in any such similar case.

Reply:
Ah, I see what you mean, you are saying that `isStartOfStmt()` returning true was wrong in the first place. @jrose-apple made the initial change to `isStartOfStmt()` so I'll defer to him for deciding if changing how `isStartOfStmt()` treats `@` tokens, and adding such assertion, is reasonable.

Reply:
I suspect that `isStartOfStmt()` returning true for `@` improves parser recovery and diagnostics for erroneously placed attributes so it's not clear to me it's a definite improvement to have it return false instead.

Reply:
It's not supposed to return true for all attributes, just those that look like statement starts.